src/System.IO.Compression/src/System/IO/Compression/Deflater.cs

```diff
@@ -169,6 +169,21 @@ internal bool Finish(byte[] outputBuffer, out int bytesRead)
             return errC == ZErrorCode.StreamEnd;
         }
 
+        internal int Flush(byte[] outputBuffer)
+        {
+            Debug.Assert(null != outputBuffer, "Can't pass in a null output buffer!");
+            Debug.Assert(NeedsInput(), "We have something left in previous input!");
+            Debug.Assert(!_inputBufferHandle.IsAllocated);
+
+            // Note: we require that NeedsInput() == true, i.e. that 0 == _zlibStream.AvailIn.
+            // If there is still input left we should never be getting here; instead we
+            // should be calling GetDeflateOutput.
+
+            int bytesRead;
+            ReadDeflateOutput(outputBuffer, ZFlushCode.SyncFlush, out bytesRead);
```

Comment:
The error checking in Deflate should be adequate to catch any unexpected errors.

ReadDeflateOutput is used by Finish, Flush, and GetDeflateOutput (Write). It will only return either ZErrorCode.Ok, ZErrorCode,StreamEnd, or ZErrorCode.BufError. Otherwise it will throw a ZlibException within the function.

For Flush and Write the possible ZErrorCode values are more limited. ZErrorCode will never be StreamEnd because StreamEnd is only occurs when the flushcode is set to ```Finish``` when calling Deflate.

So our only possible results for Flush and Write are ```ZErrorCode.BufError``` and ```ZErrorCode.Ok```. ```ZErrorCode.BufError``` is thrown when there isn't enough space in either the input buffer or output buffer to perform the Deflate. We know there will always be space in the output buffer since we're the ones passing the nonzero-length _buffer. The input buffer is a bit trickier; while the docs say that you'll get a BufError when availIn.Length == 0 calling Deflate, that is only true if there isn't anything already pending a flush. 

So the tl;dr is that it essentially comes down to whether the flush is actually necessary. If something is going to be flushed, we get ```ZErrorCode.Ok```. If there is nothing to be flushed, we get ```ZErrorCode.BufError```. 

I'm pushing an update now with a restructure of the flush methods to make use of the error code and add comments to this effect. Thanks for bringing this up.